src/joystream-lib/api.ts

```diff
@@ -0,0 +1,282 @@
+import BN from 'bn.js'
+import { ApiPromise, WsProvider } from '@polkadot/api'
+import { SubmittableExtrinsic } from '@polkadot/api/types'
+import { Bytes, GenericEvent, Raw, TypeRegistry } from '@polkadot/types'
+import { DispatchError } from '@polkadot/types/interfaces/system'
+import { web3AccountsSubscribe, web3Enable, web3FromAddress } from '@polkadot/extension-dapp'
+import { types } from '@joystream/types'
+import { ChannelCreationParameters, ContentActor, NewAsset } from '@joystream/types/content'
+import { ChannelMetadata } from '@joystream/content-metadata-protobuf'
+
+import {
+  AccountNotFoundError,
+  AccountNotSelectedError,
+  ExtensionSignCancelledError,
+  ExtensionUnknownError,
+  ExtrinsicFailedError,
+  ExtrinsicUnknownError,
+} from './errors'
+import {
+  Account,
+  AccountId,
+  AssetMetadata,
+  ChannelAssets,
+  ChannelId,
+  CreateChannelMetadata,
+  ExtrinsicStatus,
+  ExtrinsicStatusCallbackFn,
+  MemberId,
+} from './types'
+import { ContentParameters } from '@joystream/types/storage'
+import { ContentId } from '@joystream/types/media'
+
+export class JoystreamJs {
+  readonly api: ApiPromise
+  private initPolkadotExtensionPromise?: Promise<boolean>
+
+  private _selectedAccountId: AccountId | null = null
+  get selectedAccountId() {
+    return this._selectedAccountId
+  }
+
+  private _accounts: Account[] = []
+  get accounts(): Account[] {
+    return this._accounts
+  }
+
+  private unsubscribeFromAccountChanges?: () => void
+  // if needed these could become some kind of event emitter
+  public onAccountsUpdate?: (accounts: Account[]) => unknown
+  public onExtensionConnectedUpdate?: (connected: boolean) => unknown
+
+  /* Lifecycle */
+  private constructor(api: ApiPromise, appName: string) {
+    this.api = api
+
+    this.initPolkadotExtension(appName)
+  }
+
+  public async initPolkadotExtension(appName: string) {
+    this.initPolkadotExtensionPromise = this._initPolkadotExtension(appName)
+  }
+
+  private async _initPolkadotExtension(appName: string): Promise<boolean> {
+    try {
+      const enabledExtensions = await web3Enable(appName)
+
+      if (!enabledExtensions.length) {
+        this.logWarn('No extension detected')
+        this.onExtensionConnectedUpdate?.(false)
+        return false
+      }
+
+      // subscribe to changes to the accounts list
+      this.unsubscribeFromAccountChanges = await web3AccountsSubscribe((accounts) => {
+        this._accounts = accounts.map((a) => ({
+          id: a.address,
+          name: a.meta.name || 'Unnamed',
+        }))
+        this.onAccountsUpdate?.(this._accounts)
+      })
+
+      this.onExtensionConnectedUpdate?.(true)
+      return true
+    } catch (e) {
+      throw new ExtensionUnknownError(e)
+    }
+  }
+
+  static async build(appName: string, endpoint: string): Promise<JoystreamJs> {
```

Comment:
not sure what `appName` does here. Is it needed?

Reply:
yep, it's passed to `web3Enable`: https://github.com/Joystream/atlas/blob/85803b3d9814b02282a15f60d7ea8816b386a9b6/src/joystream-lib/api.ts#L65

it will be shown in the popup of the extension, asking the user if they allow connection with the app